Decorator Roadmap (#20)
lit/rfcs: rfcs/NNNN-decorator-roadmap.md

---
Status: Active
Champions: @justinfagnani

> PR: {{ update_with_pr_number }}

Author: done

This RFC describes how Lit can migrate to standard decorators with the smoothest upgrade path possible for developers.

## Objective

Migrate to using standard decorators as the only decorator implementation and unify Lit's syntax around decorators as the only way to declare reactive properties. When the migration is complete, Lit will require a decorator-supporting environment or toolchain.

### Goals
- Make Lit decorators work without transpilation
- Make the migration as easy as possible on developers
- Have a single implementation for decorators
- Unify Lit's reactive property syntax
- Simplify ReactiveElement's property handling code
- Increase component performance due to more static class initialization
- Decrease code size with simpler decorator implementations

### Non-Goals
- Change our decorators API drastically (ie, split `@property()` into `@input()`, `@output()`, `@attribute`, etc)

## Motivation

As [standard decorators](https://github.com/tc39/proposal-decorators) are starting to ship in [compilers](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#decorators) and soon in VMs, we need to prepare for migrating Lit to use them

> Standard decorators will allow us to unify our surface syntax which is currently different in plain JS and compiled sources using TypeScript or Babel. This will let us use decorators as the one way of declaring reactive properties, instead of also offering the `static properties` feature. Removing `static properties` in turn lets us remove the infrastructure for dynamically creating reactive properties, simplifying ReactiveElement.

Author: I think it's ok for that to be unsupported. They can use `Object.defineProperty` to make their own accessors, and we could give them a way to add attribute metadata if anyone comes asking for that feature.

- TypeScript developers stuck using legacy experimental decorators because of backward compatibility concerns

This plan needs to address all audiences during the migration.

### End State

After all stages of the implementation plan are complete:

- Standard decorators are the only decorator implementation
- Decorators are the one way to define reactive properties
- Remove `static properties`, `static createProperty()` and associated APIs
- Remove `static addInitializer()` since the standard decorator API provides this
- Core decorators (`@property`, `@state`) are exported from the main `reactive-element`, `lit-element` and `lit` modules.

### Standard Decorators

Standard decorators have a much different design than TypeScript's experimental decorators. Rather than getting access to the class and making dynamic, imperative changes to the class, standard decorators have no direct access to the class or the class member they're decorating and must instead return an object describing a replacement for the class member. For instance class accessor decorators return an object with get and set methods.

> This makes standard decorator usage must more static than experimental decorators. They can't add new class members, and can't change the type of member they decorate, and ReactiveElement's `static createProperty()` API is simply not callable.

Reviewer: I suspect that TS will eventually add the ability for decorators to change member types, because in practice this is something they can do at runtime

Author: Oh sorry, what I meant by "type" here is actually "kind", ie field, accessor, method, or class (and soon to be method parameter). So a decorator cannot change a field into a method, etc.

Author: changed to "kind"

> This makes standard decorator usage must more static than experimental decorators. They can't add new class members, and can't change the type of member they decorate, and ReactiveElement's `static createProperty()` API is simply not callable.

Author: fixed

The good news is that our decorator implementations are likely to be smaller and simpler. For example, the two main functions of our `@property()` decorator are to replace a field with a getter/setter pair that calls `requestUpdate()` in the setter, and to store metadata about the field for use in attribute reflection.

For example, a simplified decorator that only replaces an accessor might look like:

```ts
export const property =
  (options?: PropertyDeclaration) =>
  <C extends ReactiveElement, V>(
    _target: ClassAccessorDecoratorTarget<C, V>,
    context: ClassAccessorDecoratorContext<C, V>
  ): ClassAccessorDecoratorResult<C, V> => {
    const {access, name} = context;
    return {
      get(this: C) {
        return access.get(this);
      },
      set(this: C, v: V) {
        const oldValue = access.get(this);
        access.set(this, v);
        this.requestUpdate(name, oldValue);
      },
    };
  };
```

This decorator would require the new "auto accessor" feature and `accessor` keyword to use:

```ts
class MyElement extends LitElement {
  @property()
  accessor foo = 42;
}
```

> ## Implementation Considerations

Reviewer 2: Add detail worked out on the PR, re:
* Reflection
* Reverting to default
* ...others?

Author: done

### Implementation Plan

The plan is divided into stages which must be shipped as releases.

#### I. Add standard decorators (non-breaking)

This stage adds support for the new decorators standard in a backward-compatible way for both TypeScript and JavaScript developers.

##### Requirements

> * TypeScript ships decorator metadata

Reviewer: Worth mentioning that this is ETA TypeScript 5.2, slated for release at the end of August: https://github.com/microsoft/TypeScript/issues/54298

> ##### Changes

Reviewer 3: Will decorators imported from both `experimental-decorators.js` and `legacy-decorators.js` be possible to use together on the same class declaration?

And then when you flip to `std-decorators.js`, I assume the other imports will no longer work, and you will no longer want your TypeScript or Babel build to be configured for experimental decorators?

Author: yes, experimental-decorators and legacy-decorators will both require the `experimentalDecorators: true` tsconfig flag, and so are compatible. Standard decorators are not compatible with that flag.

* Add new module with standard decorators
    * Place them in a new module: `'lit/std-decorators.js'`
* Re-export legacy decorators from `'lit/legacy-decorators.js'`
* Add new module with a new implementation of _experimental_ decorators that work the same way standard decorators do:
    * A _third_ set of decorators: `'lit/experimental-decorators.js'`
    * They require `accessor`, replace the accessors on the prototype, do not call `static createProperty()`, store metadata in `[Symbol.metadata]`.

>     * This is so that users stuck on the experimental decorators setting (ie, google3) have a path forward when we remove the infrastructure that supports them.

Reviewer: It's generally useful for larger projects in general that want to incrementally migrate to the new decorators without multiple tsconfigs with per-file carve outs as they migrate

Reviewer 3: Great note! It's extremely useful to call out that this can be used for incremental migration.

Author: added this note

    * This decorator set will also be _use site_ compatible with standard decorators. Only the import will need to change to upgrade to standard decorators.
    * There is no great way to emulate the standard `addInitializer()`, so we keep our `static addInitializer()` un-deprecated.
* Non-core packages with decorators (@lit/labs) must follow a similar plan

>     * We can skip the legacy-decorators step with more aggressive breaking changes for labs packages.

Reviewer: Can we? context is used in google3

Reviewer: Also good chance that context will leave labs before August

Author: removed this line

> * Deprecate legacy experimental decorators, `static createProperty()`, `static getPropertyDescriptor()`, etc. but _not_ `static properties`.

Author: This should be in a separate stage. We shouldn't deprecate yet.

    * Since there are no native implementations of decorators yet, we don't quite yet want to deprecate the main developer-facing API for property declaration.
* Deprecate the `noAccessor` property option.
* Update `static elementProperties` to read from `[Symbol.metadata]`
    * Standard decorators cannot call into the `static createProperty()` API, so they must place property options into `[Symbol.metadata]`. We should use that as the source-of-truth going forward.

> * Vend codemods to upgrade decorator usage:

Reviewer 3: This codemod? https://github.com/facebookarchive/codemod

Ah or: https://github.com/facebook/jscodeshift

Regardless, it is super helpful to provide this.

    * One to migrate to the new standard decorators
    * One to migrate to the new experimental decorators

> #### II. Prefer standard decorators (breaking)

Reviewer 2: For context, might be good to add some context around how far out we expect these follow-on phases to be (iow, they're not until "xyz milestone happens, which isn't expected for xyz years"?

Author: I tried to add this with requirements, like "browsers ship native implementations". We don't know how long that will take, but hopefully we understand that browsers shipping can take years.

> This stage is still usable in non-decorator environments via `static properties`, but we move decorator modules around so that legacy use must use the non-default module names (`'lit/legacy-decorators.js'`) \

Reviewer 3: Nit: end of this line has a backslash.

Author: fixed

##### Requirements

* Stage I has landed.
* Preferred: At least one browser has shipped decorators

##### Changes

* Move standard decorators implementation to <code>'lit/decorators.js'</code>
* Re-export standard decorators from <code>'lit/std-decorators.js'</code>
* Deprecate <code>static properties</code>
* Deprecate <code>static addInitializer()</code>

> * Vend codemod to migrate standard decorator imports

Reviewer 3: This would be changing the `std-decorator.js` import to `decorators.js`?

Also, would it be worth providing a codemod to migrate legacy decorator users from `decorators.js` -> `legacy-decorators.js`, so that if a migration is taking a while, there is still a path forward?

Reviewer 3: Maybe this phase also should include a: `Deprecate lit/legacy-decorators.js` so that anyone still using those gets deprecation messages before the next phase.

Author: This includes "Deprecate legacy experimental decorators"

#### III. Remove legacy decorators (breaking)

This stage requires decorators for creating properties. It is no longer usable in non-decorator-supporting environments without transpilation.

##### Requirements

* Native decorators are shipping in all major browsers.
* Stage II has landed

##### Changes

> * Remove previously deprecated APIs

Reviewer 3: I assume this includes `'lit/legacy-decorators.js';`?

Author: yes :)

* Deprecate `'lit/std-decorators.js'` module
* Re-export `@customElement()`, `@property()` and `@state()` from the main reactive-element, lit-element, and lit modules.
    * This increases the core module size, which is paid for by removing the deprecated APIs.
    * Other decorators are more optional and remain in their own modules.

#### IV. Cleanup (breaking)

> When all developers can use the standard decorators API we can remove the experimental decorators fully.

Reviewer 3: Really niche comment, do we want to have options for https://cobalt.foo/. Or is that in the same realm as removing IE11 support in Lit 3.0?

Author: This shouldn't effect Cobal support

##### Requirements

* Native decorators are shipping in all major browsers.

> * Stage II has landed

Reviewer 3: Nit: Stage III has landed.

Author: done

- Make the migration as easy as possible on developers
- Have a single implementation for decorators
- Unify Lit's reactive property syntax
- Simplify ReactiveElement's property handling code
- Increase component performance due to more static class initialization
- Decrease code size with simpler decorator implementations

### Non-Goals

- Change our decorators API drastically (ie, split `@property()` into `@input()`, `@output()`, `@attribute`, etc)

## Motivation

As [standard decorators](https://github.com/tc39/proposal-decorators) are starting to ship in [compilers](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#decorators) and soon in VMs, we need to prepare for migrating Lit to use them.

Standard decorators will allow us to unify our surface syntax which is currently different in plain JS and compiled sources using TypeScript or Babel. This will let us use decorators as the one way of declaring reactive properties, instead of also offering the `static properties` feature. Removing `static properties` in turn lets us remove the infrastructure for dynamically creating reactive properties, simplifying ReactiveElement.

## Detailed Design

> Detailed design of the new decorators should be covered in another RFC. This RFC focuses on:

Reviewer 4: If we're going to have a separate design doc, maybe it belongs there but I don't want us to lose context on the situation described and references to the TS issues present here https://github.com/lit/lit/pull/4183#issuecomment-1711959635

Is there a place for that kind of detail here?

Reviewer 3: Took a look at a Rust RFC for inspiration https://github.com/rust-lang/rfcs/blob/master/text/0221-panic.md and I think it's totally fine to add some sentences here that link to that PR and comment. Especially so the context doesn't get lost. And this RFC is much more findable than the comment in that PR.

Reviewer 3: Decided to address this comment in the source code next to the relevant logic: https://github.com/lit/lit/pull/4336

- That standard decorators become the only way to declare reactive properties
- That breaking changes are made to remove support for dynamically adding reactive properties, since that will not be used by the new decorators
- A plan so that developers can incrementally migrate to standard decorators

### Audiences

If we group our developers by how they relate to decorators, we have four distinct audiences:

- JavaScript developers not yet using decorators
- TypeScript developers using legacy experimental decorators
- TypeScript developers not yet using decorators because of forward compatibility concerns

> - TypeScript developers stuck using legacy experimental decorators because of backward compatibility concerns

Reviewer 3: Should second point specify the legacy babel decorators?
Would the fourth option now be `JavaScript/TypeScript` developers using standard decorators.

> - TypeScript developers stuck using legacy experimental decorators because of backward compatibility concerns

Reviewer 4: I think my thinking was that "JavaScript developers using decorators with Babel" as a single group because during migration they _have_ to be using standard decorators. They have no other choice as legacy babel mode is removed.

> - TypeScript developers stuck using legacy experimental decorators because of backward compatibility concerns

Reviewer 3: Fair! I like this! I think I was coming from this retrospectively. When this RFC was written and implemented, we still supported old Babel decorators. And then chose the break them as part of Stage 1.

standard decorator usage, the `experimentalDecorators` TypeScript flag can be turned off, and semantics will remain unchanged.

#### Example migration

1. Existing codebase using experimental decorators

```ts
// tsconfig uses `experimentalDecorators: true`.
class El extends LitElement {
  @property()
  reactiveProp = "initial value";

  @state()
  secondProperty = 0;
}
```

2. Incrementally make decorator usage _use site_ compatible by adding `accessor`

```ts
// tsconfig uses `experimentalDecorators: true`.
class El extends LitElement {
  @property()
  accessor reactiveProp = "initial value";

  @state()
  accessor secondProperty = 0;
}
```

3. Finally, change `experimentalDecorators` to `false` to complete the migration.

## Implementation Considerations

### Implementation Plan

The plan is to make the Lit decorators work in both experimental decorator
environments and standard decorator environments.

#### I. Make Lit decorators hybrid (breaking)

This stage adds support for the new decorators standard for both TypeScript and JavaScript developers,
and ensures experimental decorators are semantically identical to standard decorators.

##### Requirements

> - TypeScript ships decorator metadata (TypeScript 5.2, scheduled for 2023-08-22)

Reviewer 4: ```suggestion
- TypeScript ships decorator metadata (TypeScript 5.2, scheduled for 2023-08-22)
- Babel ships decorator proposal plugin with metadata support (Babel 7.23.0, 2023-09-25)
```

##### Changes

- Make Lit decorators "hybrid", working in either experimental or standard environments.

> - Non-core packages with decorators (@lit/labs) must follow a similar plan

Reviewer 4: ```suggestion
- Non-core packages with decorators (`@lit/localize`, `@lit-labs/context`) must follow a similar plan
```

Reviewer 3: Nice!

- Deprecate the `noAccessor` property option.
- Deprecate experimental decorators, `static createProperty()`, `static getPropertyDescriptor()`, etc. but _not_ `static properties`.
  - Since there are no native implementations of decorators yet, we don't quite yet want to deprecate the main developer-facing API for property declaration.

#### III. Remove experimental decorators (breaking)

This stage requires decorators for creating properties. It is no longer usable in non-decorator-supporting environments without transpilation.

##### Requirements

- Stage II has landed
- Native decorators are shipping in all major browsers.

##### Changes

- Remove previously deprecated APIs
- Deprecate `static properties`
- Re-export `@customElement()`, `@property()` and `@state()` from the main reactive-element, lit-element, and lit modules.
  - This increases the core module size, which is paid for by removing the deprecated APIs.
  - Other decorators are more optional and remain in their own modules.

#### V. Cleanup (breaking)

When all developers can use the standard decorators API we can remove the experimental decorators fully as well as `static properties`.

##### Requirements

> - Stage IV has landed

Reviewer 4: ```suggestion
- Stage III has landed
```

> - Native decorators are shipping in all major browsers.

Reviewer 4: ```suggestion
- Native decorators have been shipping in all major browsers for some time.
```
or maybe we want to say for some specified time period/version like "for at least the last 2 major version" ?

Reviewer 3: I like the last 2 major versions verbiage. It's more concretely actionable.

Reviewer 4: ```suggestion
- Native decorators are have been shipping in all major browsers for last 2 major versions.
```

Reviewer 3: Ship it! :D

- Make the migration as easy as possible on developers
- Have a single implementation for decorators
- Unify Lit's reactive property syntax
- Simplify ReactiveElement's property handling code
- Increase component performance due to more static class initialization
- Decrease code size with simpler decorator implementations

### Non-Goals

- Change our decorators API drastically (ie, split `@property()` into `@input()`, `@output()`, `@attribute`, etc)

## Motivation

As [standard decorators](https://github.com/tc39/proposal-decorators) are starting to ship in [compilers](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#decorators) and soon in VMs, we need to prepare for migrating Lit to use them.

Standard decorators will allow us to unify our surface syntax which is currently different in plain JS and compiled sources using TypeScript or Babel. This will let us use decorators as the one way of declaring reactive properties, instead of also offering the `static properties` feature. Removing `static properties` in turn lets us remove the infrastructure for dynamically creating reactive properties, simplifying ReactiveElement.

## Detailed Design

> Detailed design of the new decorators should be covered in another RFC. This RFC focuses on:

Reviewer 3: Thank you!

- Update `static elementProperties` to read from `[Symbol.metadata]`
  - Standard decorators cannot call into the `static createProperty()` API, so they must place property options into `[Symbol.metadata]`. We should use that as the source-of-truth going forward.

#### II. Deprecate experimental decorators

##### Requirements

- Stage I has landed.
- Preferred: At least one browser has shipped decorators

##### Changes

- Deprecate the `noAccessor` property option.
- Deprecate experimental decorators, `static createProperty()`, `static getPropertyDescriptor()`, etc. but _not_ `static properties`.
  - Since there are no native implementations of decorators yet, we don't quite yet want to deprecate the main developer-facing API for property declaration.

#### III. Remove experimental decorators (breaking)

This stage requires decorators for creating properties. It is no longer usable in non-decorator-supporting environments without transpilation.

##### Requirements

- Stage II has landed
- Native decorators are shipping in all major browsers.

##### Changes

- Remove previously deprecated APIs

> - Deprecate `static properties`

Author: ```suggestion
- Deprecate `static properties` in ReactiveElement and LitElement
- Add a new mixin that supports `static properties` for migration.
```

#### II. Deprecate experimental decorators

##### Requirements

- Stage I has landed.
- Preferred: At least one browser has shipped decorators

##### Changes

- Deprecate the `noAccessor` property option.
- Deprecate experimental decorators, `static createProperty()`, `static getPropertyDescriptor()`, etc. but _not_ `static properties`.
  - Since there are no native implementations of decorators yet, we don't quite yet want to deprecate the main developer-facing API for property declaration.

#### III. Remove experimental decorators (breaking)

This stage requires decorators for creating properties. It is no longer usable in non-decorator-supporting environments without transpilation.

##### Requirements

- Stage II has landed
- Native decorators are shipping in all major browsers.

##### Changes

- Remove previously deprecated APIs
- Deprecate `static properties`
- Re-export `@customElement()`, `@property()` and `@state()` from the main reactive-element, lit-element, and lit modules.
  - This increases the core module size, which is paid for by removing the deprecated APIs.
  - Other decorators are more optional and remain in their own modules.

> #### V. Cleanup (breaking)

Reviewer 3: Note: This should be IV

Reviewer 4: For easy updating via github
```suggestion
#### IV. Cleanup (breaking)
```

When all developers can use the standard decorators API we can remove the experimental decorators fully as well as `static properties`.

##### Requirements

> 

Reviewer 3: Line beneath should be III

Reviewer 4: Fixed
